FEATURE: Handle multiple math notations

danielhollas opened the pull request:
JIRA: https://khanacademy.atlassian.net/browse/IC-662

Summary:
The current math translation system inside expects
that there is just one notation for a given operator for a given locale.
However, there might be different valid notations in some countries.
A typical example is \times and \cdot, the former is often used in lower
grades, but both are valid. We want to provide the translators the
flexibility to do both while still offering Smart Translations.

The proposed solution is a new function maybeTranslateMath(), which
decides which rules to apply by checking the template translations. I
went for the simplest heuristic I could think of. For example,
MAYBE_TIMES_AS_CDOT rule will be applied if:
1. Template translation contains \cdot at least once
anywhere in the translated text

AND

2. Template translation does not contain \times anywhere

It is unlikely that a single string would contain different notations.
If it does, we treat it as a translation mistake and ST will not be
offered.

Most importantly, this approach will allow us to handle cartesian
coordinates and intervals as well (will be done separately).
These are not possible to do atm because the US notation is the same for
both of these so we cannot determine how to translate them.

I was pleasantly surprised that the proposed change did not increase the
code complexity too much.

Obviously, this approach cannot work if the template is empty, which is
the case for strings containing only math. In this case, ST suggestions
will keep the US notation and translators will need to translate
manually if needed.

Also, we suppose that within a group, the math strings contain the same
notation, although that might not always be the case. I can imagine an
exercise that includes some problems with division and some with
multiplication, while keeping the NL text the same. In such a case, the
ST suggestions would depend on which of the string was translated as a
template. Also, this will be the case for math-only strings, where it's
quite likely to happen.
If we find this behaviour problematic, we'd need to modify the grouping
algorithm to create different groups based on used math notation.
I'd rather avoid that since it would add a lot of complexity
that I tried to avoid here. In any case, this change is a clear
improvement and should benefit translators as is.

Test Plan:
Test various locales and multiplication notations in exercise:
https://www.khanacademy.org/devadmin/content/exercises/factoring_linear_binomials/1126131850/x85f73fd8684daede

Closes #21 

(NOTE: I also implemented CDOT_AS_TIMES rule. I was assuming that US strings always use \times, but that's not the case. See #21

danielhollas commented on lib/math-translator.js:
```diff
@@ -14,14 +14,18 @@ var MATH_RULES_LOCALES = {
-    TIMES_AS_CDOT: ['cs', 'pl', 'de', 'nb', 'sr', 'ro', 'hu', 'sv', 'da', 'bg', 'lol'],
-    DIV_AS_COLON: ['cs', 'de', 'bg', 'hu', 'uk', 'da', 'hy', 'pl', 'lol', 'id', 'pt-pt', 'ru', 'nb'],
+    TIMES_AS_CDOT: ['cs', 'pl', 'de', 'nb', 'sr', 'ro', 'hu', 'sv', 'da', 'bg'],
+    CDOT_AS_TIMES: ['fr', 'ps', 'pt-pt', 'ta'],
```
This rule was missing because I previously thought that US strings only contain `times`.
So this is not strictly related to this PR, but I included it so that I can update all locale rules accordingly and have complete treatment for multiplication notation.

danielhollas commented on src/math-translator.js:
```diff
@@ -15,16 +15,21 @@ const MATH_RULES_LOCALES = {
-            'lol'],
-    DIV_AS_COLON: ['cs', 'de', 'bg', 'hu', 'uk', 'da', 'hy', 'pl', 'lol', 'id',
+    TIMES_AS_CDOT: ['cs', 'pl', 'de', 'nb', 'sr', 'ro', 'hu', 'sv', 'da', 'bg'],
+    CDOT_AS_TIMES: ['fr', 'ps', 'pt-pt', 'ta'],
```
This rule was missing because I previously thought that US strings only contain times.
This is not strictly related to this PR, but I included it so that I can update all locale rules accordingly and have a complete treatment for multiplication notation.

mathrobo replied:
Not a direct comment on \\times and \\cdot, but was thinking about alternate structures like \\div and \\frac{}{} (Would that be more complex since it's not simply symbol replacement? Just pure intellectual curiosity, not actual need.)

danielhollas replied:
Yeah, this would be doable albeit a bit harder to handle all possible corner cases. But I think for this particular case it's not needed because at least in the primary school, div and frac are different concepts, so you wouldn't want to mix them up.

danielhollas commented on src/math-translator.js:
```diff
@@ -15,16 +15,21 @@ const MATH_RULES_LOCALES = {
     ARABIC_COMMA: ['ps'],
     PERSO_ARABIC_NUMERALS: ['ps'],
+    MAYBE_DIV_AS_COLON: ['id', 'lol'],
+    MAYBE_TIMES_AS_CDOT: ['az', 'bn', 'el', 'hi', 'hy', 'id', 'it', 'ja', 'ka',
```
In principle, MAYBE_TIMES_AS_CDOT === MAYBE_CDOT_AS_TIMES, but I made it as a separate rule to preserve symmetry with other rules. But I do not have a strong opinion on this.

danielhollas replied:
I added a test to make sure they are equal.

danielhollas commented on src/math-translator.js:
```diff
@@ -240,14 +249,50 @@ function normalizeTranslatedMath(math, lang) {
     return math;
 }
 
-const MathTranslator = {};
```
This was a remnant from previous PR.

danielhollas commented on src/translation-assistant.js:
```diff
@@ -138,16 +139,23 @@ function getMapping(
+            // even though in principle a single string can very well contain
+            // both an interval (0,10) and cartesian point (0, 0) for example
+            // Presumably, such cases are rare.
+            .map((input) => maybeTranslateMath(input, outputs.join(' '), lang))
```
in this case, `outputs` hold translated math bits from a translator and we're trying to match the notation that they used.

danielhollas replied:
Yeah, unfortunately, this is a chicken-or-egg problem, we're in the `getMapping()` function that creates this mapping: translators can change the order of math bits. I guess we could be more creative, but this is the simplest solution. Also, we do not want to support mixed notation within a single string anyways so I don't think this is an issue.

mpolyak replied:
Thank you for the explanation Daniel. Could you please add in a comment as well, e.g.:

```
.map((input) =>
    // We pass in the entire template...
    maybeTranslateMath(input, outputs.join(' '), lang))
```

Better yet maybe extract `outputs.join(' ')` above, e.g.:
```
// We provide the entire template...
const translatedTemplate = outputs.join(' '); // Not sure about the name?

inputs = inputs
    .map((input) => translateMath(input, lang))
    .map((input) => maybeTranslateMath(input, translatedTemplate, lang))
```

?

danielhollas replied:
Okay, thanks! I've called it `allTranslatedMaths` and did the same in all three instances where we call `maybeTranslateMath`. I've also switched the order: `maybeTranslateMath` should be called before the numbers are localized in `translateMath`. It doesn't matter here, but will be important for coordinates and intervals.

danielhollas commented on src/translation-assistant.js:
```diff
@@ -138,16 +139,23 @@ function getMapping(
+        // NOTE(danielhollas): Smart Translations will not be offered
+        // if translator did not translate math according to our locale rules,
+        // normalizeTranslatedMath only handles some special cases
+        // where we permit variations
```
Just rewording a note to make it clearer.

danielhollas replied:
Done, thanks!

danielhollas commented:
@mpolyak this one is ready for review as well. Sorry for a somewhat lengthy writeup. :blush:

danielhollas commented on src/math-translator.js:
```diff
@@ -240,14 +249,50 @@ function normalizeTranslatedMath(math, lang) {
 
-const MathTranslator = {};
+/**
+ * We will translate certain math notation
+ * to match the notation in the translated template string
+ * this is needed for cases where two or more translations are valid
+ * (\cdot versus \times)
+ * or where there's not a unique mapping from US notation
+ * (intervals and cartesian coordinates have the same notation in the US
+ * but different in many countries).
```
Done, thanks! I also added an example.

danielhollas commented on src/math-translator.js:
```diff
@@ -240,14 +249,50 @@ function normalizeTranslatedMath(math, lang) {
+    const maybeMathTranslations = [
+        // multiplication sign as a centered dot
+        {langs: MATH_RULES_LOCALES.MAYBE_TIMES_AS_CDOT,
+            regex: /\\times/g, replaceRegex: /\\cdot/, replace: '\\cdot'},
```
Yes, I was not happy with the naming, `fromRegex` and `toRegex` are great!

Unfortunately, `toRegex` will not match `replace` for intervals and coordinates, where we will have for example (the actual regex will be more complex):

```js
fromRegex: /\((\d+), (\d+)\)/,
toRegex: /\[\d+, \d+\]/,
replace: '[$1, $2]'
```

So the code looks silly now but will be needed in the next PR.

danielhollas replied:
Actually, I realized that the intervals/coordinates will need to be handled separately anyway, so I've simplified this and used the same syntax as in `translateMath()`

danielhollas commented on tests/assistant_spec.js:
```diff
@@ -553,6 +565,43 @@ describe('TranslationAssistant (math-translate)', function() {
         assertSuggestions(allItems, itemsToTranslate, translatedStrs, 'de');
     });
 
+    it('should translate math notations based on template', function() {
+        const allItems = [{
+            englishStr: 'simplify $3\\times x = 9.9 \\div 3$',
+            translatedStr: 'simplifyz $3\\cdot x = 9{,}9 \\mathbin{:} 3$',
+        }];
+        const itemsToTranslate = [
+            {englishStr: 'simplify $5\\times x=0.9\\div 3$', translatedStr: ''},
+        ];
+        const translatedStrs = ['simplifyz $5\\cdot x=0{,}9\\mathbin{:} 3$'];
+        assertSuggestions(allItems, itemsToTranslate, translatedStrs, 'lol');
+    });
```
I think it would not pass because `lol` lang is in `MAYBE_TIMES_AS_CDOT` and not in `TIMES_AS_CDOT`. But you're right this test is not very instructive, I've modified it and added more tests that will hopefully make that clearer.

danielhollas commented on tests/math_translator_spec.js:
```diff
@@ -231,3 +241,72 @@ describe('MathTranslator (normalizeTranslatedMath)', function() {
+describe('MathTranslator (maybeTranslateMath)', function() {
+    it('should translate times to cdot according to template', function() {
+        MATH_RULES_LOCALES.MAYBE_TIMES_AS_CDOT.forEach(function(locale) {
+            const englishStr = '8 \\times 2 = 16';
+            const template1 = '8 \\cdot 1 = 8';
+            const template2 = '8 \\times 1 = 8';
+            const translatedStr1 = '8 \\cdot 2 = 16';
+            const translatedStr2 = '8 \\times 2 = 16';
+            const output1 = maybeTranslateMath(englishStr, template1, locale);
+            const output2 = maybeTranslateMath(englishStr, template2, locale);
+            assert.equal(output1, translatedStr1);
+            assert.equal(output2, translatedStr2);
```
Yep, done. The first is not really different from Translate math, but the point is that `maybeTranslateMath` only applies the rules that it detects in the template, wheras `translateMath` applies all the rules unconditionally.

danielhollas commented on src/translation-assistant.js:
```diff
@@ -589,6 +594,16 @@ class TranslationAssistant {
             }
 
+            // Translate items that are only math even if there's no template.
+            // TODO(kevinb) handle multiple non-nl_text items
```
I've separated and moved this code below the code that translates strings according to a template so that `maybeTranslateMath` can be applied to math-only strings. If the template is not available, we'll apply only `translateMath()` here.

danielhollas commented on tests/assistant_spec.js:
```diff
@@ -254,6 +255,24 @@ describe('TranslationAssistant', function() {
             assertSuggestions(allItems, itemsToTranslate, translatedStr);
         });
 
+        it('should return the same graphie even if translator changed one',
```
This is testing an existing behavior.

danielhollas commented on tests/assistant_spec.js:
```diff
@@ -268,6 +287,24 @@ describe('TranslationAssistant', function() {
             assertSuggestions(allItems, itemsToTranslate, translatedStr);
         });
+
+        it('should return the same imageLink even if translator changed one',
```
Again, just ensuring the existing behavior.

danielhollas commented:
@mpolyak this is now ready for a new review. :tada: 

Besides implementing your suggestions, I added a lot of tests, fixed behavior for strings containing `\text{}' and/or `\textbf`, and improved treatment of strings containing only math.

This got a bit complex, so I've pulled these changes to `hollas-manticore` branch on webapp. It might be useful if you could deploy that to ZND so we can test it in the real world. Thanks!

danielhollas commented on tests/math_translator_spec.js:
```diff
@@ -285,3 +295,142 @@ describe('MathTranslator (normalizeTranslatedMath)', function() {
+        });
+    });
+
+    it('should not have conflicting rules for trig functions', function() {
```
I'll add more such tests in a future PR.

mpolyak commented on src/math-translator.js:
```diff
@@ -284,14 +294,64 @@ function normalizeTranslatedMath(math, lang) {
+ * that the translator used in the template string.
+ *
+ * This is needed for cases where two or more translations are valid,
+ * e.g. \cdot versus \times), or where there's not a unique mapping from
```
typo in "\times)"

mpolyak commented on src/math-translator.js:
```diff
@@ -284,14 +294,64 @@ function normalizeTranslatedMath(math, lang) {
+            !template.match(el.fromRegex)) {
+
+            math = math.replace(el.fromRegex, el.replace);
+
```
nit: blank link

danielhollas replied:
Hm, I actually like the extra blank line here to separate the if statement from the code, the indentation would not be super clear here.

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -831,6 +868,157 @@ describe('TranslationAssistant (math-translate)', function() {
+describe('TranslationAssistant (maybe-math-translate)', function() {
+
+    it('should translate math according to a template', function() {
+        // Lang 'id' is in both MAYBE_TIMES_AS_CDOT and MAYBE_DIV_AS_COLON
```
You could also add an assert to that effect, in case it's ever removed, it'll be more obvious why the test breaks.

`assert(MATH_RULES_LOCALES.MAYBE_TIMES_AS_CDOT.includes(lang))`?

danielhollas replied:
Good idea, I've done it for this first test.

mpolyak commented on tests/assistant_spec.js:
```diff
@@ -831,6 +868,157 @@ describe('TranslationAssistant (math-translate)', function() {
+        assertSuggestions(allItems, itemsToTranslate, translatedStrs, lang);
+    });
+
+    it('should return the same math without a template', function() {
```
I wonder if `it.each` is possible:

```
it.each(
    MATH_RULES_LOCALES.MAYBE_TIMES_AS_CDOT
)("should return the same math without a template for %s", function(lang) {

});
```

danielhollas replied:
Apparently not. :-( We could iterate over locales inside the callback in the same way we do in `math_translater_spec`, but I'll keep it as is for now. I'll try to clean up the tests in some future PR. A lot of them could be made clearer if the lang was chosen more explicitly (e.g. via asserting that lang is in certain rules as you suggested above).

danielhollas commented:
@mpolyak thanks for the deploy! By the way, if I pushed new commits to the branch, would they be automatically picked up or would you need to re-deploy?

I tested on ZND and it seems to do what I think it should be doing.

 I've made some final cleanup and adjustments and this should be good to go. :tada:  :crossed_fingers:

danielhollas commented:
Awesome, thank you!

I'm tracking the overall work in this Jira:
https://khanacademy.atlassian.net/browse/CP-3984
I do not have access to Phabricator, but I am not sure if that's intentional or not. I'll ask Caroline.

I think it'll take a couple of weeks before we get there though. I have two more PRs that I want to do, and I need to ask Advocates for input to update the rules.
